Feature/Cities CRUD

FelipeCasalG opened the pull request:
# Feature/Cities CRUD

## Description ⭐

<!--
Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.
Example:
-->

I've worked on the cities model, migrations, controller and routing to establish the api with the CRUD methods.
The changes also include some basic components that will be used across the blade frontend and the necessary views to work with the cities.

I did the cities index view with blade php. This page fetches information from the backend via Ajax. It has pagination, sorting and does not reload on every request. 

### Type of change 🏃

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)
- [ ] This change requires a documentation update

### Requires ❗

- [x] This change requires a database update or migration
- [ ] This change requires a documentation update

## How Has This Been Tested? ❓

<!--
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
-->

No unit testing added to the project yet

FelipeCasalG committed:
feat: Update homepage to use a dashboard layout

Introduce a set of components and a layout with tailwind css to start working on the UI

FelipeCasalG committed:
feat: Add City model, migration, factory, and seeder

This commit adds the City model, migration, factory, and seeder files.

FelipeCasalG committed:
feat: Add CityController with CRUD operations for cities

This commit adds a new controller called CityController with CRUD (Create, Read, Update, Delete) operations for managing cities. It also includes the necessary routes in api.php and the setup of the api in the app.php file.

FelipeCasalG committed:
feat: Update CityController to use PATCH method for updating cities

FelipeCasalG committed:
feat: Add index view for cities

This commit generates the index view for the cities. As part of the work, I've created some blade components that will be reused in other views in the future

FelipeCasalG committed:
feat: Remove unnecesary comment in CityController

danielaWaldeck95 commented:
Feli comentario sobre el Pull Request. Buenisimo que usaste el template!
Cositas a tener en cuenta:
- El titulo que sea representativo de tu PR
- Todo lo que no se use que hace ruido visual borralo del md. Por ejemplo los checkboxes que no aplican, si no tiene task en clickup el link se puede borrar, la seccion de screenshots si no incluye etc etc..

Otra cosa, trata de hacer Pull Requests mas pequeños siempre que pueda separarse en pequeñas features

danielaWaldeck95 commented on app/Http/Controllers/CityController.php:
```diff
@@ -0,0 +1,81 @@
+        return response()->json(['cities' => $cities]);
+    }
+
+    public function store(Request $request): JsonResponse
```
You could use FormRequest to define all the rules. Then you dont need to care about validaror->fails() for example. 

You can read more about FormRequests here https://laravel.com/docs/11.x/validation#form-request-validation

danielaWaldeck95 commented on app/Http/Controllers/CityController.php:
```diff
@@ -0,0 +1,81 @@
+        $validator = Validator::make(
+            $request->all(),
+            [
+                'name' => 'required|string|max:255|unique:cities,name',
```
```suggestion
                'name' => ['required', 'string', 'max:255', 'unique:cities,name'],
```

This will be in FormRequest rules but anyways we prefer to write rules as an array instead of using | as separation

danielaWaldeck95 commented on app/Http/Controllers/CityController.php:
```diff
@@ -0,0 +1,81 @@
+
+        $city->name = $request->input('name');
+        $city->save();
+        return response()->json($city, 200);
```
You could use Flugg\Responder and define a transformer to respond with the created city

danielaWaldeck95 commented on app/Http/Controllers/CityController.php:
```diff
@@ -0,0 +1,81 @@
+
+    public function destroy($id): JsonResponse
+    {
+        $city = City::find($id);
```
You could use FirstOrFail for example and It will fail if no city is found with that id. Then catch the exception.

FelipeCasalG replied:
I used 
```
        catch (ModelNotFoundException $e) {
            return response()->json(['error' => 'City not found'], JsonResponse::HTTP_NOT_FOUND);
        }
```

danielaWaldeck95 commented on routes/api.php:
```diff
@@ -0,0 +1,9 @@
+use Illuminate\Support\Facades\Route;
+use App\Http\Controllers\CityController;
+
+Route::get('/cities', [CityController::class, 'getAll']);
+Route::post('/cities', [CityController::class, 'store']);
+Route::patch('/cities/{id}', [CityController::class, 'update']);
+Route::delete('/cities/{id}', [CityController::class, 'destroy']);
```
Remember to use one controller per endpoint with an execute method

danielaWaldeck95 commented on routes/web.php:
```diff
@@ -1,7 +1,8 @@
-});
+Route::view('/', 'dashboard');
+
+Route::get('/cities', [CityController::class, 'index']);
```
idem

danielaWaldeck95 commented on app/Http/Controllers/CityController.php:
```diff
@@ -0,0 +1,81 @@
+            'name' => $request->input('name'),
+        ]);
+
+        return response()->json($city, 201);
```
Avoid hardcoding response codes for example for 201 you could use `JsonResponse::HTTP_CREATED`

danielaWaldeck95 commented on app/Http/Controllers/CityController.php:
```diff
@@ -0,0 +1,81 @@
+        }
+
+        $city = City::create([
+            'name' => $request->input('name'),
```
since the input is a string you can do this 
`$request->string('name')->toString()`

FelipeCasalG committed:
feat: Update CityController and city index view for pagination

This commit updates the CityController to use pagination when fetching cities.

The cities index page was updated with the pagination logic.

Additionally, the CitySeeder has been updated to create 35 cities instead of 10 for testing purposes.

FelipeCasalG committed:
fix: fixed composer fixer errors and open PR comments

FelipeCasalG committed:
feat: added the create city feature in the cities page

Also changed the controller to use form requests

FelipeCasalG committed:
refactor: use one controller per endpoint with execute method

FelipeCasalG committed:
refactor: added laravel responder to the project

Updated all controllers with responder

FelipeCasalG committed:
feat: added sorting to get all cities endpoint

FelipeCasalG committed:
feat: Handle invalid query parameters in GetAllCitiesController

FelipeCasalG committed:
fet: added sorting to view cities

FelipeCasalG committed:
fix: phpstan and fixer errors solved

FelipeCasalG committed:
fix: solved routing issues with __invoke methods

FelipeCasalG committed:
fix: solved issues with Form Request return type

Added failedValidation function to solve it

FelipeCasalG committed:
Composer automatic fixes

FelipeCasalG committed:
fix: removed unnecesary failedValidation

Found a better workaround by adding the header Accept/Application json in postman

danielaWaldeck95 commented on routes/api.php:
```diff
@@ -0,0 +1,12 @@
+use App\Http\Controllers\GetAllCitiesController;
+use App\Http\Controllers\UpdateCityController;
+
+Route::get('/cities', GetAllCitiesController::class);
+Route::post('/cities', CreateCityController::class);
+Route::patch('/cities/{id}', UpdateCityController::class);
+Route::delete('/cities/{id}', DestroyCityController::class);
```
you could add a name for each route and then use that name instead of hardcoding the url in your views 
```suggestion
Route::get('/cities', GetAllCitiesController::class)->name('cities');
Route::post('/cities', CreateCityController::class)->name('cities.store');
Route::patch('/cities/{id}', UpdateCityController::class)->name('cities.update');
Route::delete('/cities/{id}', DestroyCityController::class)->name('cities.destroy');
```

danielaWaldeck95 replied:
also to avoid repetition of /cities you could use a prefix and group all your routes by the same prefix. Check it here: https://laravel.com/docs/11.x/routing#route-group-prefixes

This will be also useful when working with middleware and authorization.

danielaWaldeck95 commented on resources/views/cities/index.blade.php:
```diff
@@ -0,0 +1,167 @@
+    let sort = null;
+    let order = null;
+    function fetchCities(page) {
+        url = '/api/cities?page=' + page;
```
Here you could use your route by its name

danielaWaldeck95 commented on resources/views/cities/index.blade.php:
```diff
@@ -0,0 +1,167 @@
+
+                // Update pagination links
+                $('#pagination').html('');
+                if (response.pagination) {
+                    $('#pagination').append(
+                        '<nav class="flex items-center justify-between border-t border-gray-200 px-4 py-3 sm:px-6"\
+                                aria-label="Pagination">\
+                                <div class="hidden sm:block">\
+                                    <p class="text-sm text-gray-700">Showing page<span class="font-medium"> ' + response.pagination.currentPage + ' \
+                                        </span>of<span class="font-medium"> ' + response.pagination.totalPages + ' \
+                                        </span></p>\
+                                </div>\
+                                <div id="paginationBtn" class="flex flex-1 justify-between sm:justify-end">\
+                                </div>\
+                            </nav>'
+                    );
+                    if (response.pagination.currentPage > 1) {
+                        $('#paginationBtn').append(
+                            '<a onclick="fetchCities(' + (response.pagination.currentPage - 1) +
+                            ')"\
+                                class="relative inline-flex items-center rounded-md px-3 py-2 text-sm font-semibold text-gray-900 ring-1 ring-inset ring-gray-300 hover:bg-gray-200 focus-visible:outline-offset-0">Previous</a>'
+                        );
+                    } else {
+                        $('#paginationBtn').append(
+                            '<a class="relative ml-3 inline-flex items-center rounded-md px-3 py-2 text-sm font-semibold text-gray-400 ring-1 ring-inset ring-gray-200 focus-visible:outline-offset-0 cursor-not-allowed">Previous</a>'
+                        );
+                    }
+                    if (response.pagination.currentPage < response.pagination.totalPages) {
+                        $('#paginationBtn').append(
+                            '<a onclick="fetchCities(' + (response.pagination.currentPage + 1) +
+                            ')"\
+                                class="relative ml-3 inline-flex items-center rounded-md px-3 py-2 text-sm font-semibold text-gray-900 ring-1 ring-inset ring-gray-300 hover:bg-gray-200 focus-visible:outline-offset-0">Next</a>'
+                        );
+                    } else {
+                        $('#paginationBtn').append(
+                            '<a class="relative ml-3 inline-flex items-center rounded-md px-3 py-2 text-sm font-semibold text-gray-400 ring-1 ring-inset ring-gray-200 focus-visible:outline-offset-0 cursor-not-allowed">Next</a>'
+                        );
```
What happens if you have an other paginated response in the future? You'd have to add all this logic again? Try something more reusable, check laravel documentation on displaying pagination here: https://laravel.com/docs/11.x/pagination#displaying-pagination-results
and also on customising the display editing the vendor here: https://laravel.com/docs/11.x/pagination#customizing-the-pagination-view

danielaWaldeck95 commented on resources/views/cities/index.blade.php:
```diff
@@ -0,0 +1,167 @@
+        }
+        $.ajax({
+            type: 'GET',
+            url: url,
```
you could use data: to add params to your query instead of concatenating strings to your url

danielaWaldeck95 commented on app/Models/City.php:
```diff
@@ -0,0 +1,22 @@
+
+class City extends Model
+{
+    protected $fillable = ['name'];
```
Here you could use `protected $guarded = ['id'];`  instead of fillable.
That way you don't have to add every attribute you want to add to your model to the fillable array. Its easier to keep updated the guarded array.

FelipeCasalG committed:
fix: Solved PR comments

FelipeCasalG committed:
fix: solving phpstan errors

danielaWaldeck95 commented on database/seeders/DatabaseSeeder.php:
```diff
@@ -15,9 +15,11 @@ public function run(): void
-            'name' => 'Test User',
-            'email' => 'test@example.com',
-        ]);
+        // User::factory()->create([
+        //     'name' => 'Test User',
+        //     'email' => 'test@example.com',
+        // ]);
+
+        $this->call(CitySeeder::class);
```
remove commented code

FelipeCasalG committed:
fix: Remove unused comments from code in DatabaseSeeder.php
